lib/dnscache.c

```diff
@@ -469,7 +471,12 @@ dns_caching_thread_deinit(void)
   g_assert(dns_cache != NULL);
 
   G_LOCK(unused_dns_caches);
-  unused_dns_caches = g_list_prepend(unused_dns_caches, dns_cache);
+
+  if (main_loop_is_terminating(main_loop_get_instance()))
+    dns_cache_free(dns_cache);
+  else
```

Comment:
hmm I don't really like this coupling between dnscache and mainloop. Why isn't the dns_caching_global_deinit() enough? That should cover the exit case properly.

Reply:
As mainloop does not wait threads to terminate, it can happen that `unused_dns_caches` is freed by the mainloop in `dns_caching_global_deinit` by the time` dns_caching_thread_deinit` is called. This results in an invalid read in the thread:
```
unused_dns_caches = g_list_prepend(unused_dns_caches, dns_cache);
```
as `unused_dns_caches` is already freed.

The leak here is only a side effect, due that `dns_cache` is appended to a non existing list.

Reply:
ugh, I understand. Still I would leave in this as a bug and then fix the termination issue in mainloop perhaps later. 

As far as I understand, the mainloop does have a grab on how many threads are executing, and we could wait for that to drop to 0, perhaps even using a timeout, to ensure that it does not take all eternity to exit.

If the timeout is reached, we would exit drastically (e.g. exit() without calling the shutdown functions), but that should only happen in extreme cases.

Reply:
Ok, fine for me. I drop this patch and try to create a separate one based on your guidance.